Support config defaults using .pip-tools.toml or pyproject.toml

j00bar opened the pull request:
Supports use of a `.pip-tools.toml`, `pyproject.toml`, or other explicitly specified TOML file to set configuration defaults for `pip-compile` and `pip-sync`.

Fixes #604.

##### Contributor checklist

- [X] Provided the tests for the changes.
- [X] Assure PR title is short, clear, and good to be included in the user-oriented changelog

##### Maintainer checklist

- [ ] Assure one of these labels is present: `backwards incompatible`, `feature`, `enhancement`, `deprecation`, `bug`, `dependency`, `docs` or `skip-changelog` as they determine changelog listing.
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

webknjaz commented:
@j00bar you can't trigger CI until you have at least one commit in the repo..

webknjaz commented on piptools/scripts/compile.py:
```diff
@@ -302,6 +303,20 @@ def _determine_linesep(
+        allow_dash=False,
+        path_type=str,
+    ),
+    help="Path to a pyproject.toml file with specialized defaults for pip-tools",
```
Per my comments in the issue, it is important to allow a pip-tools specific config taking preference over `pyproject.toml`. Also, the way it's implemented, it'll allow naming the file arbitrarily, so this isn't accurate with regards to the generic case, only the default. Let's try to rephrase this without implying that this filename is mandatory.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +525,104 @@ def parse_requirements_from_wheel_metadata(
         )
+
+
+def pyproject_toml_defaults_cb(
```
This needs to be renamed for the same reason. It is inaccurate that this only reads `pyproject.toml`.
Also, since it's a function, its name should start with a verb describing the action it performs, not a contextless static state.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +525,104 @@ def parse_requirements_from_wheel_metadata(
+    ctx: click.Context, param: click.Parameter, value: str | None
+) -> str | None:
+    """
+    Defaults for `click.Command` parameters should be override-able in pyproject.toml
```
Could you make the docstring title describe what the function does following PEP 257?

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +525,104 @@ def parse_requirements_from_wheel_metadata(
+
+
+def mutate_option_to_click_dest(option_name: str) -> str:
+    "Mutates an option from how click/pyproject.toml expect them to the click `dest` value"
```
Use triple double backticks as customary for docstrings, imperative mood, short title, long paragraph separately.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +525,104 @@ def parse_requirements_from_wheel_metadata(
+}
+
+
+def mutate_option_to_click_dest(option_name: str) -> str:
```
This name is misleading, I don't see the implementation mutating the incoming object, it seems to return a new one.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +525,104 @@ def parse_requirements_from_wheel_metadata(
+
+
+@functools.lru_cache()
+def parse_pyproject_toml(config_file: str) -> dict[str, Any]:
```
Again, let's not stick references to `pyproject.toml` all over the place. Looks like this just reads a tool config, regardless of the filename.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on pyproject.toml:
```diff
@@ -42,6 +42,7 @@ dependencies = [
   "pip >= 22.2",
   # indirect dependencies
   "setuptools", # typically needed when pip-tools invokes setup.py
+  "toml >= 0.10.1",
```
This line is out of place, the section says “indirect deps” but it's actually used directly now.

Though, this is not the only problem here: it should also be conditional — stdlib features a toml parser since Python 3.11 so it doesn't need an extra dep to work.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on tests/test_utils.py:
```diff
@@ -540,3 +543,102 @@ def test_get_sys_path_for_python_executable():
+
+def test_pyproject_toml_defaults_cb_bad_toml(make_pyproject_toml_conf):
+    config_file = make_pyproject_toml_conf("verbose", True)
+    config_text = open(config_file).read()
+    open(config_file, "w").write(config_text[::-1])
```
This opens 2 file descriptors and never closes them. It will cause `ResourceWarning`s that are usually turned into errors in pytest when in strict mode.
It's best to use pathlib for the inline invocations like this. Otherwise, use context managers.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on tests/test_utils.py:
```diff
@@ -540,3 +543,102 @@ def test_get_sys_path_for_python_executable():
+    open(config_file, "w").write(config_text[::-1])
+    ctx = Context(compile_cli)
+    ctx.params["src_files"] = (str(config_file),)
+    pytest.raises(FileError, pyproject_toml_defaults_cb, ctx, "config", None)
```
Please, use a context manager. Also, it is a best practice to always use the `match=` argument with such test helpers.

webknjaz commented on tests/test_utils.py:
```diff
@@ -540,3 +543,102 @@ def test_get_sys_path_for_python_executable():
+
+def test_pyproject_toml_defaults_cb_unreadable_toml(make_pyproject_toml_conf):
+    ctx = Context(compile_cli)
+    pytest.raises(
```
Same as above, plz use a CM.

webknjaz replied:
Marking this thread as resolved.

webknjaz reviewed (changes requested):
Thanks for working on this! I was able to skim through the patch from mobile and noticed a few areas that might need some love...
Though, I'm sure someone will check the logic too. Reviews are easier from an actual computer.

I'll also approve the CI in just a bit to see how it's doing...

webknjaz commented:
> I'll also approve the CI in just a bit to see how it's doing...

The PyPy failures seem unrelated... Though, they'll probably need sorting out separately before this PR could be mergeable. I wonder if bumping the interpreter version would have any effect.

j00bar commented:
Thank you for the detailed feedback!

1. I've added support and preference for a `.pip-tools.toml` file.
2. I've implemented the best practices and conventions you referenced.
3. I've made the dependency on `toml` apply only pre-3.11 - the tests still require the external toml library, as `tomllib` reads only but does not yet write.
4. I've updated the README to document the new configurability.

webknjaz commented on README.rst:
```diff
@@ -306,6 +306,29 @@ Any valid ``pip`` flags or arguments may be passed on with ``pip-compile``'s
 -------------
 
+You can define project-level defaults for ``pip-compile`` and ``pip-sync`` by
+writing them to a configuration file in the same directory as your requirements
```
Has this changed since the previous review? I think I saw CWD being used in a few places, not an input-related location...

j00bar replied:
It had not, no. If you specify `src_files`, it looks in the directories of the `src_files` - if you do not specify `src_files`, it uses `cwd()`.

So `pip-compile /path/to/requirements.in` would look for `/path/to/.pip-tools.toml` or `/absolute/path/to/pyproject.toml` -- whereas `pip-compile` without arguments would look in the current working directory.

At first blush, that sounds like the behavior that would be expected, versus looking in the current directory for those files, even when compiling a source in another directory. Happy to revise if you disagree.

webknjaz replied:
I don't disagree, just wanted to clarify that I understand this correctly.

webknjaz replied:
I think this paragraph should mention the corner case of what's happening when the input is coming from stdin. Also, should we fall back to the output file directory when we read from stdin but the output is specified via a CLI option?

webknjaz replied:
Fair enough. Marking as resolved.

webknjaz commented on README.rst:
```diff
@@ -306,6 +306,29 @@ Any valid ``pip`` flags or arguments may be passed on with ``pip-compile``'s
+.. code-block:: toml
+
+    # In a .pip-tools.toml file
+    [pip-tools]
```
I'm unsure about this. The section name could probably be the same as for `pyproject.toml` which would make it easier to migrate between the files... OTOH, `tool.` might be excessive here.
```suggestion
    [tool.pip-tools]
```
This is not a request to make the change right away but rather an idea to be discussed.

j00bar replied:
Certainly - happy to have that discussion. I was comparing with other projects and trying for consistency. You can see, for example, that in this project's `pyproject.toml` that mypy, setuptools, setuptools_scm, and pytest all have the `tool` prefix - whereas flake8 and bandit's bespoke files just use the name of the tool as the section of the file.

webknjaz replied:
Yeah, I mostly just wanted to bring attention to this aspect. You wanted to ensure the implementation remains simple, and I saw a possibility to make that happen.

On a more general note, I understand that the tools listed come with a lot of legacy and their tool-specific configs don't use TOML at all.

I have a precedent for you, though. Towncrier, for example, expects `[tool.towncrier]` in both of its supported config file locations — `towncrier.toml` and `pyproject.toml`. And this makes sense — what if an end-user passes a config file path with an arbitrary name of `my-cool-custom-config.toml` — would we expect the section to be `[pip-tools]` or `[tool.pip-tools]`?

AndydeCleyre replied:
I notice that for [ruff](https://beta.ruff.rs/docs/configuration/#using-rufftoml), when using a `ruff.toml`/`.ruff.toml`, the settings can either be under `[tool.ruff]`, or without a top level heading at all.

webknjaz replied:
My original motivation was to help simplify the patch and the code base in general. Adding a conditional would keep some of the complexity. Do we really want that in order to save people one line? I'm conflicted here. Though, it's not as important to me so I won't be insisting of a specific resolution of this.

j00bar replied:
I think my recommendation would be to maintain consistency with the rest of the ecosystem insofar as we can ascertain it. I would accept an incremental degree of code complexity to achieve a better developer experience with minimal cognitive load.

Surveying popular tools in the Python ecosystem, it seems that `[tool.pip-tools]` is the convention for TOML files and `[pip-tools]` would be the convention for INI-style files. I'll make it consistent, since we're using TOML everywhere.

webknjaz replied:
Agreed. Marking this thread as resolved.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
+    if not config:
+        return None
+
+    defaults: dict[str, Any] = ctx.default_map.copy() if ctx.default_map else {}
+    defaults.update(config)
+
+    ctx.default_map = defaults
```
Why is this needed? Would the following be equivalent?
```suggestion
    if ctx.default_map is None:
        ctx.default_map = {}
    ctx.default_map.update(config)
```

FWIW, pieces like this could benefit from being in a dedicated function with a descriptive name.

j00bar replied:
I didn't have an explicit reason for it - I was following a suggestion in the PR to look at the way black implemented this as a click callback and used their code as an inspiration, in which they reconstruct a whole new dict for `default_map` instead of updating the current one. The `git blame` seems to indicate the suggested change you've offered would run afoul of the type-checker.

https://github.com/psf/black/blame/b2ee211b5ad84b62738ac0997b73bf6ee9a74d06/src/black/__init__.py#L129-L134C1

webknjaz replied:
Honestly, I wouldn't recommend doing whatever black does, as it has countless shortcomings. To me, it's rather an indication that an idea might be bad.
I understand that in-browser code suggestions often need to be corrected later on, locally. This is still a good way of visualizing the expectations/ideas/suggestions...

webknjaz replied:
OTOH, I was making the suggestions based on the engineering thinking rather than whether some other project does something in one way or another. Keeping a project easy-to-read/follow and maintainable sometimes requires more attention to detail. As I was progressing with the review, my perception, and understanding changed/became more accurate, which is why I added new suggestions on top of the previous ones. Making pauses helped me gain better perspective..

webknjaz replied:
I'll try to see if that 3-year old problem got solved by now..

webknjaz replied:
> The `git blame` seems to indicate the suggested change you've offered would run afoul of the type-checker.

Good news — mypy didn't explode on this.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
+    defaults.update(config)
+
+    ctx.default_map = defaults
+    return str(config_file)
```
Operating on strings for paths is a bad API. Use pathlib wherever possible, except for places interfacing with external things that only understand strings (like printing out to logs/console).

webknjaz replied:
```suggestion
    return config_file
```

webknjaz commented on pyproject.toml:
```diff
@@ -40,6 +40,7 @@ dependencies = [
   "build",
   "click >= 8",
   "pip >= 22.2",
+  "toml >= 0.10.1; python_version<'3.11'",
```
Let's make it a little bit more sparse, like the first part.
```suggestion
  "toml >= 0.10.1; python_version < '3.11'",
```

webknjaz replied:
with the recent comments in the Python modules, this should be
```suggestion
  "tomli >= 0.10.1; python_version < '3.11'",
```

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
+
+@functools.lru_cache()
+def parse_config_file(config_file: Path) -> dict[str, Any]:
+    if sys.version_info >= (3, 11):
+        # Python 3.11 stdlib tomllib load() requires a binary file object
+        with config_file.open("rb") as ifs:
+            config = tomllib.load(ifs)
+    else:
+        # Before 3.11, using the external toml library, load requires the filename
+        config = toml.load(str(config_file))
```
Having seen this, I remembered that CPython specifically chose to mimic the popular library APIs. I also recall that all the major ecosystem projects moving away from some toml libraries. Namely, `toml` was something that stopped being used, along with `tomlkit`. I think, I still have one project using `toml` that I have to migrate.
I've verified that the library you should be using is called `tomli`. I checked that it's now used by `pip`, `pypa/build`, `pytest` and `tox`, among others.

Once you fix that, it'll be possible to make this somewhat more elegant:
```suggestion
    config = tomllib.loads(config_file.read_text(encoding="utf-8"))
```

webknjaz commented on piptools/utils.py:
```diff
@@ -2,14 +2,22 @@
+import sys
+from pathlib import Path
 from typing import TYPE_CHECKING, Any, Callable, Iterable, Iterator, TypeVar, cast
 
 import click
+
+if sys.version_info >= (3, 11):
+    import tomllib
+else:
+    import toml
```
```suggestion

if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib

import click
```

webknjaz commented on pyproject.toml:
```diff
@@ -57,6 +58,7 @@ testing = [
   "pytest >= 7.2.0",
   "pytest-rerunfailures",
   "pytest-xdist",
+  "toml >= 0.10.1",
```
```suggestion
  "tomli-w",
```

webknjaz commented on piptools/utils.py:
```diff
@@ -405,9 +414,9 @@ def get_required_pip_specification() -> SpecifierSet:
+    requirement = next(
         (r for r in project_dist.requires() if r.name == "pip"), None
-    )
+    )  # pragma: no branch
```
This doesn't seem related to this PR.

j00bar replied:
This appears to be an artifact of pre-commit reorganizing code.

webknjaz replied:
That's interesting. I wonder why this didn't happen in the PRs bumping the hooks.. Anyway, I'll mark this thread as resolved, it's not a blocker.

webknjaz commented on piptools/scripts/compile.py:
```diff
@@ -340,6 +356,7 @@ def cli(
     emit_index_url: bool,
     emit_options: bool,
     unsafe_package: tuple[str, ...],
+    config: str | None,
```
Let's make this accept an already normalized `pathlib.Path` object:
```suggestion
    config: Path | None,
```

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
+    else:
+        config_file = Path(value)
+
+    try:
+        config = parse_config_file(config_file)
+    except OSError as e:
+        raise click.FileError(
+            filename=str(config_file), hint=f"Could not read '{config_file}': {e}"
+        )
+    except ValueError as e:
+        raise click.FileError(
+            filename=str(config_file), hint=f"Could not parse '{config_file}': {e}"
+        )
+
+    if not config:
+        return None
+
+    defaults: dict[str, Any] = ctx.default_map.copy() if ctx.default_map else {}
+    defaults.update(config)
+
+    ctx.default_map = defaults
```
Upon looking into this more, I'm now convinced that this does not belong here at all. It should happen externally, rather than being tightly coupled to the path argument validation and normalization. Being a side effect only makes the control flow more entangled.

j00bar replied:
I think this bears more discussion.

The order of precedence in determining the value for a particular option is: 1. An explicit declaration at runtime via a flag; 2. A default specified in a config file; 3. A default hard-coded in pip-tools. The configuration file's values need to override the built-in defaults but not any explicitly declared values. By the time the `cli()` function is called, explicit declarations and hard-coded defaults have already been resolved with one another, and it becomes convoluted to try to determine for which options the config file's values should be operative. The same holds true for decorators.

By using an `is_eager=True` option with a callback, we have the opportunity to selectively override hard-coded defaults with new config-file-specified defaults before any commandline explicit options are handled. I don't see a simple place to inject this order of precedence otherwise, and from Click documentation, that follows the intent of the `is_eager` flag:

> Sometimes, you want a parameter to completely change the execution flow. [ ... ] In such cases, you need two concepts: eager parameters and a callback.

If you can imagine an approach using a decorator or in the `cli()` function that can nimbly achieve this order of precedence, I'm open to hearing it.

webknjaz replied:
Okay, now I understand the challenges better. Marking this thread as resolved.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
+    config_file_path = next(
+        (
+            candidate_dir / config_file
+            for candidate_dir in candidate_dirs
+            for config_file in (CONFIG_FILE_NAME, "pyproject.toml")
```
It might be nice to make use of `itertools.product()` here.

webknjaz replied:
This would probably be an overkill, marking this thread as resolved.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
+        candidate_dirs = [Path.cwd()]
+    else:
+        # Collect the candidate directories based on the src_file arguments provided
+        src_files_as_paths = [
+            Path(Path.cwd(), src_file).resolve() for src_file in src_files
+        ]
+        candidate_dirs = [
+            src if src.is_dir() else src.parent for src in src_files_as_paths
+        ]
```
Let's make these lazy
```suggestion
        src_files_as_paths = (
            Path(Path.cwd(), src_file).resolve() for src_file in src_files
        )
        candidate_dirs = (
            src if src.is_dir() else src.parent for src in src_files_as_paths
        )
```

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
         )
+
+
+def callback_config_file_defaults(
```
After looking into what this function doers, it became clear why it's so hard to call it properly — it does too much and has side effects.

This should probably be a simple input-to-normalized path pure function.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
+def callback_config_file_defaults(
+    ctx: click.Context, param: click.Parameter, value: str | None
+) -> str | None:
+    """
+    Returns the path to the config file with defaults being used, or `None` if no such file is
+    found.
```
Let's not forget that the docstring title should be in imperative mood, short and simple, referencing to the action being performed, not describing the entire runtime logic. PEP 8 requires such indented blocks not to exceed the limit of 72 chars:
```suggestion
    """Return the config file path.

    ``None`` is returned if no such file is found.
```
(also, we should only use RST in docstrings)

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
+    else:
+        # Collect the candidate directories based on the src_file arguments provided
+        src_files_as_paths = [
+            Path(Path.cwd(), src_file).resolve() for src_file in src_files
```
It took me a minute to get why you're feeding Path into Path... How about
```suggestion
            (Path.cwd() / src_file).resolve() for src_file in src_files
```

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
+]
+
+
+@functools.lru_cache()
```
nit: does this need to be called if you don't pass any args?
```suggestion
@functools.lru_cache
```

webknjaz replied:
OTOH, maybe it makes sense to limit the cache size...
```suggestion
@functools.lru_cache(max_size=1)
```

webknjaz commented on tests/conftest.py:
```diff
@@ -450,3 +453,21 @@ def _reset_log():
+
+
+@pytest.fixture
+def make_config_file(tmpdir_cwd):
+    def _maker(pyproject_param, new_default, config_file_name=CONFIG_FILE_NAME):
```
could you try adding typing  here?

webknjaz commented on tests/conftest.py:
```diff
@@ -450,3 +453,21 @@ def _reset_log():
+
+
+@pytest.fixture
+def make_config_file(tmpdir_cwd):
```
Definitely add a docsting — it's shown in the `pytest --fixtures` output and is a part of good DX/contributore experience.

webknjaz replied:
Marking this as resolved.

webknjaz commented on tests/conftest.py:
```diff
@@ -450,3 +453,21 @@ def _reset_log():
+        config_to_dump = {"pip-tools": {pyproject_param: new_default}}
+        if config_file_name == "pyproject.toml":
+            config_to_dump = {"tool": config_to_dump}
+        with open(config_file, "w") as ofs:
+            toml.dump(config_to_dump, ofs)
+        return config_file
```
with tomli, it's
```suggestion
        config_file.write_text(tomli_w.dumps(config_to_dump))
        return config_file
```

webknjaz commented on tests/conftest.py:
```diff
@@ -8,9 +8,11 @@
 from textwrap import dedent
 
 import pytest
+import toml
```
```suggestion
import tomli_w
```

webknjaz commented:
@j00bar I apologize that this all came off quite overwhelming — my reviews are often like this when I get excited, which causes people to think I'm too nitpicky... This is a collaboratively-maintained project, meaning that hundreds of people have maintenance privileges. Though, only a few maintain specific Jazzband projects consistently. I'm reviewing this because I care about certain aspects, plus you pinged me. It could've remained unreviewed for some while otherwise. Of course, there's always a chance somebody would step in to give extra feedback.

Some bits I commented are optional, but nice to have. Others stand firm ground. In particular, that decoupling request.
You mentioned that it'changing it would require rewriting tests substantially but I don't see why — could you clarify what you foresee? I must've missed some implications, eh?

Finally, there's no time pressure to iterate on feedback. But since you mentioned that you'd postpone further commits, would it be okay if I committed something to your branch myself to ease the burden?

webknjaz commented on .pre-commit-config.yaml:
```diff
@@ -33,6 +33,7 @@ repos:
           - toml==0.10.2
           - pip==20.3.4
           - build==0.9.0
+          - types-toml==0.10.2
```
This will no longer be relevant with `tomli`
```suggestion
```

webknjaz commented on piptools/scripts/compile.py:
```diff
@@ -19,13 +19,14 @@
 from .._compat import parse_requirements
```
```suggestion
from pathlib import Path
from .._compat import parse_requirements
```

webknjaz commented on piptools/scripts/compile.py:
```diff
@@ -391,6 +408,9 @@ def cli(
         )
 
+    if config:
+        log.info(f"Using pip-tools configuration defaults found in '{config}'.")
```
```suggestion
        log.info(f"Using pip-tools configuration defaults found in '{config !s}'.")
```

webknjaz commented on piptools/scripts/sync.py:
```diff
@@ -103,6 +120,7 @@ def cli(
     client_cert: str | None,
     src_files: tuple[str, ...],
     pip_args: str | None,
+    config: str | None,
```
```suggestion
    config: Path | None,
```

webknjaz commented on piptools/scripts/sync.py:
```diff
@@ -127,6 +145,9 @@ def cli(
             sys.exit(2)
 
+    if config:
+        log.info(f"Using pip-tools configuration defaults found in '{config}'.")
```
```suggestion
        log.info(f"Using pip-tools configuration defaults found in '{config !s}'.")
```

webknjaz commented on piptools/scripts/sync.py:
```diff
@@ -17,9 +17,11 @@
 from .._compat import parse_requirements
```
```suggestion
from pathlib import Path
from .._compat import parse_requirements
```

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,120 @@ def parse_requirements_from_wheel_metadata(
+
+def callback_config_file_defaults(
+    ctx: click.Context, param: click.Parameter, value: str | None
+) -> str | None:
```
```suggestion
) -> Path | None:
```

webknjaz commented on tests/test_utils.py:
```diff
@@ -540,3 +544,113 @@ def test_get_sys_path_for_python_executable():
+    ctx = Context(compile_cli)
+    ctx.params["src_files"] = (str(config_file),)
+    found_config_file = callback_config_file_defaults(ctx, "config", None)
+    assert found_config_file == str(config_file)
```
```suggestion
    assert found_config_file == config_file
```

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,129 @@ def parse_requirements_from_wheel_metadata(
+    if config := parse_config_file(config_file):
+        _assign_config_to_cli_context(ctx, config)
+    else:
+        return None
```
@j00bar through a series of code conversions I discovered that the check here is rather weird — `if (not) config:` checks for a falsy value (which a present but empty section would be).
This seems to be semantically incorrect. If there's an empty config, we'd still want to recognize its existence.
Is this the reason you thought it'd be hard to decouple this?
Why do we even want to pretend a config file wasn't passed at all to the CLI when we don't see a config there? I don't think we should do that..

j00bar replied:
You're correct - thank you.

webknjaz replied:
@j00bar looks like this behavior hasn't been corrected.

webknjaz commented on tests/test_utils.py:
```diff
@@ -540,3 +544,113 @@ def test_get_sys_path_for_python_executable():
+    ctx.params["src_files"] = (str(project_config_file),)
+    found_config_file = callback_config_file_defaults(ctx, "config", None)
+    # The pip-tools specific config file should take precedence over pyproject.toml
+    assert found_config_file == str(piptools_config_file)
```
```suggestion
    assert found_config_file == piptools_config_file
```

webknjaz commented on piptools/scripts/compile.py:
```diff
@@ -19,13 +20,14 @@
 from ..resolver import BacktrackingResolver, LegacyResolver
 from ..utils import (
     UNSAFE_PACKAGES,
+    callback_config_file_defaults,
```
```suggestion
    determine_config_file,
```

webknjaz commented on piptools/scripts/compile.py:
```diff
@@ -302,6 +304,21 @@ def _determine_linesep(
+    help=f"Read configuration from TOML file. By default, looks for a {CONFIG_FILE_NAME} or "
+    "pyproject.toml.",
+    is_eager=True,
+    callback=callback_config_file_defaults,
```
```suggestion
    callback=determine_config_file,
```

webknjaz commented on piptools/scripts/sync.py:
```diff
@@ -17,9 +18,11 @@
 from ..logging import log
 from ..repositories import PyPIRepository
 from ..utils import (
+    callback_config_file_defaults,
```
```suggestion
    determine_config_file,
```

webknjaz commented on piptools/scripts/sync.py:
```diff
@@ -86,6 +89,21 @@
+    help=f"Read configuration from TOML file. By default, looks for a {CONFIG_FILE_NAME} or "
+    "pyproject.toml.",
+    is_eager=True,
+    callback=callback_config_file_defaults,
```
```suggestion
    callback=determine_config_file,
```

webknjaz commented on tests/test_utils.py:
```diff
@@ -5,19 +5,23 @@
 from piptools.scripts.compile import cli as compile_cli
 from piptools.utils import (
     as_tuple,
+    callback_config_file_defaults,
```
```suggestion
    determine_config_file,
```

webknjaz commented on tests/test_utils.py:
```diff
@@ -540,3 +544,113 @@ def test_get_sys_path_for_python_executable():
+    # Create a "compile" run example pointing to the config file
+    ctx = Context(compile_cli)
+    ctx.params["src_files"] = (str(config_file),)
+    found_config_file = callback_config_file_defaults(ctx, "config", None)
```
```suggestion
    found_config_file = determine_config_file(ctx, "config", None)
```

webknjaz commented on tests/test_utils.py:
```diff
@@ -540,3 +544,113 @@ def test_get_sys_path_for_python_executable():
+    ctx = Context(compile_cli)
+    ctx.params["src_files"] = (str(config_file),)
+    with pytest.raises(BadOptionUsage, match="must be a list"):
+        callback_config_file_defaults(ctx, "config", None)
```
```suggestion
        determine_config_file(ctx, "config", None)
```

webknjaz commented on tests/test_utils.py:
```diff
@@ -540,3 +544,113 @@ def test_get_sys_path_for_python_executable():
+    ctx = Context(compile_cli)
+    ctx.params["src_files"] = (str(config_file),)
+    with pytest.raises(FileError, match="Could not parse "):
+        callback_config_file_defaults(ctx, "config", None)
```
```suggestion
        determine_config_file(ctx, "config", None)
```

webknjaz commented on tests/test_utils.py:
```diff
@@ -540,3 +544,113 @@ def test_get_sys_path_for_python_executable():
+    project_config_file = make_config_file("newline", "CRLF", "pyproject.toml")
+    ctx = Context(compile_cli)
+    ctx.params["src_files"] = (str(project_config_file),)
+    found_config_file = callback_config_file_defaults(ctx, "config", None)
```
```suggestion
    found_config_file = determine_config_file(ctx, "config", None)
```

webknjaz commented on tests/test_utils.py:
```diff
@@ -540,3 +544,113 @@ def test_get_sys_path_for_python_executable():
+def test_callback_config_file_defaults_unreadable_toml(make_config_file):
+    ctx = Context(compile_cli)
+    with pytest.raises(FileError, match="Could not read "):
+        callback_config_file_defaults(
```
```suggestion
        determine_config_file(
```

q0w commented on piptools/utils.py:
```diff
@@ -522,3 +531,129 @@ def parse_requirements_from_wheel_metadata(
+    else:
+        config_file = Path(value)
+
+    if config := parse_config_file(config_file):
```
py37 is already dropped?

webknjaz replied:
For some reason I thought we already have access to walrus... Python 3.7 is going EOL in 1.5 months so we'll have to wait.
```suggestion
    config = parse_config_file(config_file)
    if config:
```

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,130 @@ def parse_requirements_from_wheel_metadata(
+    """
+    Returns the config file to use for defaults given `src_files` provided.
+    """
+    if not src_files:
```
I've found a way to reduce branching here. Will add a commit for this.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented:
@j00bar I've addressed some of the comments and I'm making a pause now. Maybe I'll get back to this later, or maybe you will. There's some extra discussions that happened FYI, please check them out and respond as you have time.

chrysle commented on piptools/scripts/compile.py:
```diff
@@ -391,6 +409,9 @@ def cli(
         )
 
+    if config:
+        log.info(f"Using pip-tools configuration defaults found in '{config !s}'.")
```
This line should get a test.

chrysle commented on piptools/scripts/sync.py:
```diff
@@ -127,6 +146,9 @@ def cli(
             sys.exit(2)
 
+    if config:
+        log.info(f"Using pip-tools configuration defaults found in '{config !s}'.")
```
Same here.

webknjaz replied:
@j00bar it'd be nice to know why isn't this covered..

webknjaz commented:
@chrysle creating merge commits in PR branches brings contributors pain. It's best not to add them for the sake or adding, especially since there's clearly more things to do here and this causes unnecessary disruptions on shared branches like this.

chrysle commented:
Apologies @webknjaz 

I wanted to fix the RTD build. I will remember this for the future.

webknjaz commented:
> I wanted to fix the RTD build.

It'd be fixed on the next commit by itself, I think. Most CI-like systems use the ephemeral merge commit of PR branches into main which combines states of both.

If anything, I'd go for using the rebase setting of the branch update, but only if the PR was about to be merged so that the contributing experience isn't messed up by force pushes.

webknjaz commented on tests/test_utils.py:
```diff
@@ -540,3 +544,113 @@ def test_get_sys_path_for_python_executable():
+    ),
+)
+def test_callback_config_file_defaults(pyproject_param, new_default, make_config_file):
+    config_file = make_config_file(pyproject_param, new_default)
```
The need to inject this invocation into every test could be eliminated by making the params indirect (this needs the respective fixtures to depend on `request` and read `request.param`).

j00bar replied:
It's a clever idea, and if I we were making the config file based on the key and value being passed in parametrized values in multiple tests, I'd agree it may be worth the effort. But this is the only invocation that gets all of its arguments from test parameters - the others either have some or all that are constants. Given that, indirect parameterization feels like over-engineering that makes the test scenario more opaque. Thoughts?

webknjaz replied:
Okay, this one not as important. Marking as resolved for now.

webknjaz commented on tests/test_utils.py:
```diff
@@ -540,3 +544,113 @@ def test_get_sys_path_for_python_executable():
+        determine_config_file(
+            ctx,
+            "config",
+            "/path/does/not/exist/my-config.toml",
```
This would probably be a better non-existence guarantee:
```suggestion
            "/dev/null/path/does/not/exist/my-config.toml",
```

webknjaz commented:
@j00bar I'll need to dedicate more time to reviewing this, but there's one pattern I noticed that has to be fixed — your docstrings use single backticks for things that should be marked as inline code. In RST, unlike MD, these must be double backticks. The single ones have different semantics. There's currently no sane way to have MD in docstrings and RST is the only thing virtually any Sphinx-based project is able to consume.

webknjaz commented:
@j00bar looks like you accidentally re-reverted [`4980a8f` (#1863)](https://github.com/jazzband/pip-tools/pull/1863/commits/4980a8f23a94e9fd67afefbfcd7f768acf048cdb). Could you re-apply/cherry-pick this commit yet again for Python 3.7 compat?

webknjaz commented:
Thanks for the updates! It may take me some time to review the PR — I got ill 🤒 and my brain is rather defunct atm..
Though, if I'm not answering for too long, feel free to ping me. TIA!

j00bar commented:
> I got ill 🤒

Damn. Feel better, friend. The PR will wait.

webknjaz commented:
Thanks, y'all! I'm mostly fine now, so I'll try to review the updates.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
+        config_file = Path(value)
+
+    config = parse_config_file(config_file)
+    if config:
```
Per https://github.com/jazzband/pip-tools/pull/1863#discussion_r1217283171, this should probably be
```suggestion
    if config is not None:
```

webknjaz replied:
@j00bar I think this is the last bit that needs clearing up... And maybe a test?

j00bar replied:
I don't think this suggestion makes sense anymore - in a [prior commit](https://github.com/jazzband/pip-tools/pull/1863/commits/93507f9c021d4b23d0379c988c80ed5fc999a8e4#diff-4d72add98acf8c426866899357a936987e567ad30d6b9064c97d94ea96cd77baL558-L559) I removed an `else: return None`, which was the problem, as you said: an empty configuration file shouldn't be treated as if there were no configuration file found/specified.

`config`'s value comes from `parse_config_file`, which returns a `dict[str, Any]`, so `config` would never be None. And since `_assign_config_to_cli_context` basically does a `default_map.update()` using the contents of `config`, with an empty dict that would be a no-op, so this conditional saves us that call.

webknjaz replied:
Alright, I see. Marking this thread as resolved.

webknjaz replied:
It would be nice to have a regression test with this corner case but it's not a blocker.

webknjaz commented:
@j00bar with the rst to md conversion, there's a conflict with the readme now that needs solving.

webknjaz commented on piptools/utils.py:
```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
         )
+
+
+def determine_config_file(
```
Yes, love it! Marking this as resolved.

j00bar commented:
Rebased and translated the README changes to Markdown. :+1:

webknjaz reviewed (approved):
I hereby declare this worthy! Excellent work!

Any additional concerns can be addressed in follow-ups.

AndydeCleyre commented:
With this, how does one ensure no arguments from configuration files interfere with what's manually run (and visible) on the command line? Do we need to specify `--config /dev/null`, if that works, to prevent surprises, and ensure consistent behavior with the current release?

AndydeCleyre commented:
Also, are the config-file defaults merged with what's explicitly passed on the command line, or entirely replaced if any options (or args?) are passed?

j00bar commented:
@AndydeCleyre The order of precedence, from highest to lowest:

* Explicit command-line parameter
* Config file
* Hard coded defaults

So if you specify something on the command-line, it overrides whatever is in the config file. If you specify something in the config file, it overrides whatever is hard coded into pip-tools.

For settings with multiple values, there's no merging. It follows the order of precedence.

Does that make sense?

AndydeCleyre commented:
That makes sense but it doesn't clear up the questions for me.

1. If I have the option `cert` specified in the config file, and I pass `--annotation-style=line` on the command line, will the `cert` option be used?
2. If I want to run `pip-compile` with the program's defaults, even if there exists a hidden `.pip-tools.toml` file in the current directory (i.e. I want it to act as `pip-compile` does in the current release), is it now necessary that I change all current calls of arg-less `pip-compile` to `pip-compile --config /dev/null`? Will that work? Is that the only way to achieve this?

webknjaz commented:
These are great points. It's probably worth adding said test cases to make the expectations explicit. It might be nice to document this too and implement a `--no-config`..

j00bar commented:
> 1. If I have the option `cert` specified in the config file, and I pass `--annotation-style=line` on the command line, will the `cert` option be used?

Yes. Each parameter follows the order of precedence independently. The full configuration, like most command-line utilities, uses a mosaic of command-line set arguments, config file arguments, and hardcoded defaults.

> > 2. If I want to run `pip-compile` with the program's defaults, even if there exists a hidden `.pip-tools.toml` file in the current directory (i.e. I want it to act as `pip-compile` does in the current release), is it now necessary that I change all current calls of arg-less `pip-compile` to `pip-compile --config /dev/null`? Will that work? Is that the only way to achieve this?

>  It might be nice to... implement a --no-config

I don't agree. Projects use these config files to set defaults that should apply to the whole project to make the preferred way of compiling/syncing requirements be the easiest way to compile/sync requirements. That encourages consistency transparently. As the steward of many projects and having to have created automation to fail a PR when it fails consistency in the compiled requirements, I want it as laborious as possible to deviate from the project's default settings. And if there are deviations I often make, I'm going to make an additional config file with a non-default config file name.

But you're a maintainer of this one - it's up to you if you want that - but I'm not interested in helping with that one.

AndydeCleyre commented:
> @AndydeCleyre I wonder why is this backwards incompatible?

I still don't know the answer to this question, but this is what I was thinking of:

> If I want to run `pip-compile` with the program's defaults, even if there exists a hidden `.pip-tools.toml` file in the current directory (i.e. I want it to act as `pip-compile` does in the current release), is it now necessary that I change all current calls of arg-less `pip-compile` to `pip-compile --config /dev/null`? Will that work? Is that the only way to achieve this?

atugushev commented:
> I still don't know the answer to this question

@AndydeCleyre 

I'm inclined to think that this update does not represent a breaking change. The current update does not interfere with pip-compile's default behavior, nor does it require any modifications to existing calls. Running pip-compile without any arguments will still function as it does in the current release and shouldn't break any existing pipelines, as there were no configurations before.

However, I believe we should consider adding support for `--no-config` (similar to how it's done in wget or ripgrep) in the upcoming version and resolve other concerns.

Are there any objections if I remove the corresponding label and proceed with the version 6.14.0 (#1892) release process?

AndydeCleyre commented:
Ok, remove the breaking label and status. 

I would really like to know how to consistently protect myself from unwanted configs once this is released. 

And yes, that absolute path is going to be a paper cut for any teams that include the header and run pip-compile on different systems. IMO all paths in the txt output should be relative to the txt file.

atugushev commented:
@AndydeCleyre thanks!

I've prepared two PRs to address [my](https://github.com/jazzband/pip-tools/pull/1863#issuecomment-1611728125):
-  https://github.com/jazzband/pip-tools/pull/1893
- https://github.com/jazzband/pip-tools/pull/1894


Let's get this merged before release.

atugushev commented:
Released as part of pip-tools v6.14.0.

webknjaz commented:
Hey @j00bar, FYI there's some post-release adjustment and bugfixing efforts going on, in case you wanted to keep an eye on the changes, here's some:
- https://github.com/jazzband/pip-tools/pull/1893
- https://github.com/jazzband/pip-tools/pull/1896
- https://github.com/jazzband/pip-tools/issues/1902
- https://github.com/jazzband/pip-tools/issues/1903
- https://github.com/jazzband/pip-tools/pull/1904

AndydeCleyre commented:
When using `.pip-tools.toml`, does there need to be a `[tool.pip-tools]` section, or should the values be top-level with no section header? Whichever the case, can it be noted somehow in the README?

chrysle commented:
@AndydeCleyre 

> When using `.pip-tools.toml`, does there need to be a `[tool.pip-tools]` section, or should the values be top-level with no section header? 

See these lines:

https://github.com/jazzband/pip-tools/blob/776c3fa3d8b2da155eed7eb9a5b164f78937987a/piptools/utils.py#L656-L657

Maybe this could be simplified for `.pip-tools.toml`.

> Whichever the case, can it be noted somehow in the README?

I think it's explained properly under [the configuration section](https://github.com/jazzband/pip-tools#configuration).

j00bar commented:
> Maybe this could be simplified for .pip-tools.toml.

That was the original implementation which, in review, was rejected in favor of consistency.

AndydeCleyre commented:
Thanks. I thought there was a discussion about making it like ruff, but I can't find it so maybe dreamed it. From ruff's docs:

> As an alternative to `pyproject.toml`, Ruff will also respect a `ruff.toml` (or `.ruff.toml`) file, which implements an equivalent schema (though in the `ruff.toml` and `.ruff.toml` versions, the `[tool.ruff]` header is omitted).

webknjaz commented:
> Thanks. I thought there was a discussion about making it like ruff, but I can't find it so maybe dreamed it. From ruff's docs:
> 
> > As an alternative to `pyproject.toml`, Ruff will also respect a `ruff.toml` (or `.ruff.toml`) file, which implements an equivalent schema (though in the `ruff.toml` and `.ruff.toml` versions, the `[tool.ruff]` header is omitted).

FTR, that discussion that you're recalling was about the need to have the section header in all of the files or not. It was you who mentioned Ruff in that context: https://github.com/jazzband/pip-tools/pull/1863#discussion_r1187752607.

And I made an argument that it's more consistent to have this header in all files: https://github.com/jazzband/pip-tools/pull/1863#discussion_r1188761535. It makes it easier to copy the section over between different files. And now I think that when the end-users want to have several sections in their non-pyproject config, they can only do so when there are sections in the file.
